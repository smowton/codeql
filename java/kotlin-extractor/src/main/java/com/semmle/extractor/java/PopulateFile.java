package com.semmle.extractor.java;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

import com.github.codeql.Label;
import com.github.codeql.TrapWriter;
import com.github.codeql.KotlinExtractorDbSchemeKt;
import com.semmle.util.trap.pathtransformers.PathTransformer;
import com.semmle.util.files.FileUtil;

// A temporary port of TrapWriter.populateFile et al, until we can just use the existing TrapWriter or split it up

public class PopulateFile {

  private TrapWriter tw;
  private PathTransformer transformer;
  public PopulateFile(TrapWriter tw) {
    this.tw = tw;
    this.transformer = PathTransformer.std();
  }

	private static final String[] keyReplacementMap = new String[127];
	static {
		keyReplacementMap['&'] = "&amp;";
		keyReplacementMap['{'] = "&lbrace;";
		keyReplacementMap['}'] = "&rbrace;";
		keyReplacementMap['"'] = "&quot;";
		keyReplacementMap['@'] = "&commat;";
		keyReplacementMap['#'] = "&num;";
	}

	/**
	 * Escape a string for use in a TRAP key, by replacing special characters with HTML entities.
	 * <p>
	 * The given string cannot contain any sub-keys, as the delimiters <code>{</code> and <code>}</code>
	 * are escaped.
	 * <p>
	 * To construct a key containing both sub-keys and arbitrary input data, escape the individual parts of
	 * the key rather than the key as a whole, for example:
	 * <pre>
	 * "foo;{" + label.toString() + "};" + escapeKey(data)
	 * </pre>
	 */
	public static String escapeKey(String s) {
		StringBuilder sb = null;
		int lastIndex = 0;
		for (int i = 0; i < s.length(); ++i) {
			char ch = s.charAt(i);
			switch (ch) {
			case '&':
			case '{':
			case '}':
			case '"':
			case '@':
			case '#':
				if (sb == null) {
					sb = new StringBuilder();
				}
				sb.append(s, lastIndex, i);
				sb.append(keyReplacementMap[ch]);
				lastIndex = i + 1;
				break;
			}
		}
		if (sb != null) {
			sb.append(s, lastIndex, s.length());
			return sb.toString();
		} else {
			return s;
		}
	}

  public Label populateFile(File absoluteFile) {
    String databasePath = transformer.fileAsDatabaseString(absoluteFile);
    // Ensure the rewritten path is used from now on.
    File normalisedFile = new File(databasePath);
    Label result = tw.getLabelFor("@\"" + escapeKey(databasePath) + ";sourcefile" + "\"");
    KotlinExtractorDbSchemeKt.writeFiles(tw, result, databasePath);
    populateParents(normalisedFile, result);
    return result;
  }

	private Label addFolderTuple(String databasePath) {
		Label result = tw.getLabelFor("@\"" + escapeKey(databasePath) + ";folder" + "\"");
		KotlinExtractorDbSchemeKt.writeFolders(tw, result, databasePath);
		return result;
	}

	/**
	 * Populate the parents of an already-normalised file. The path transformers
	 * and canonicalisation of {@link PathTransformer#fileAsDatabaseString(File)} will not be
	 * re-applied to this, so it should only be called after proper normalisation
	 * has happened. It will fill in all parent folders in the current TRAP file.
	 */
	private void populateParents(File normalisedFile, Label label) {
		File parent = normalisedFile.getParentFile();
		if (parent == null) return;

    Label parentLabel = addFolderTuple(FileUtil.normalisePath(parent.getPath()));
    populateParents(parent, parentLabel);
		KotlinExtractorDbSchemeKt.writeContainerparent(tw, parentLabel, label);
	}

}