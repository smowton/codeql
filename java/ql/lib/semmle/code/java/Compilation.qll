/**
 * Provides a class representing individual compiler invocations that occurred during the build.
 */

import semmle.code.FileSystem

/**
 * An invocation of the compiler. Note that more than one file may be
 * compiled per invocation. For example, this command compiles three
 * source files:
 *
 *   javac Foo.java Bar.java Baz.java
 *
 * Two things happen to each file during a compilation:
 *
 *   1. The file is compiled by a real compiler, such as javac or kotlinc.
 *   2. The file is parsed by the CodeQL front-end.
 *   3. The parsed representation is converted to database tables by
 *      the CodeQL extractor.
 *
 * This class provides CPU and elapsed time information for steps 2 and 3,
 * but not for step 1.
 */
class Compilation extends @compilation {
  /** Gets a textual representation of this element. */
  string toString() {
    exists(string name |
      compilations(this, _, _, name) and
      result = "<compilation " + name + ">"
    )
  }

  /** Gets a file compiled during this invocation. */
  File getAFileCompiled() { result = getFileCompiled(_) }

  /** Gets the `i`th file compiled during this invocation */
  File getFileCompiled(int i) { compilation_compiling_files(this, i, result) }

  /**
   * Gets the amount of CPU time spent processing file number `i` in the
   * front-end.
   */
  float getFrontendCpuSeconds(int i) { compilation_time(this, i, 1, result) }

  /**
   * Gets the amount of elapsed time while processing file number `i` in the
   * front-end.
   */
  float getFrontendElapsedSeconds(int i) { compilation_time(this, i, 2, result) }

  /**
   * Gets the amount of CPU time spent processing file number `i` in the
   * extractor.
   */
  float getExtractorCpuSeconds(int i) { compilation_time(this, i, 3, result) }

  /**
   * Gets the amount of elapsed time while processing file number `i` in the
   * extractor.
   */
  float getExtractorElapsedSeconds(int i) { compilation_time(this, i, 4, result) }

  /**
   * Gets an argument passed to the extractor on this invocation.
   */
  string getAnArgument() { result = getArgument(_) }

  /**
   * Gets the `i`th argument passed to the extractor on this invocation.
   */
  string getArgument(int i) { compilation_args(this, i, result) }

  /**
   * Gets the total amount of CPU time spent processing all the files in the
   * compiler.
   */
  float getCompilerCpuSeconds() { compilation_compiler_times(this, result, _) }

  /**
   * Gets the total amount of elapsed time while processing all the files in
   * the compiler.
   */
  float getCompilerElapsedSeconds() { compilation_compiler_times(this, _, result) }

  /**
   * Gets the total amount of CPU time spent processing all the files in the
   * front-end and extractor.
   */
  float getTotalCpuSeconds() { compilation_finished(this, result, _) }

  /**
   * Gets the total amount of elapsed time while processing all the files in
   * the front-end and extractor.
   */
  float getTotalElapsedSeconds() { compilation_finished(this, _, result) }

  /**
   * Holds if this is a compilation of Java code.
   */
  predicate isJava() { this instanceof @javacompilation }

  /**
   * Holds if this is a compilation of Kotlin code.
   */
  predicate isKotlin() { this instanceof @kotlincompilation }

  /**
   * Holds if extraction for the compilation started.
   */
  predicate extractionStarted() { compilation_started(this) }

  /**
   * Holds if the extractor terminated normally. Terminating with an exit
   * code indicating that an error occurred is considered normal
   * termination, but crashing due to something like a segfault is not.
   */
  predicate normalTermination() { compilation_finished(this, _, _) }
}
