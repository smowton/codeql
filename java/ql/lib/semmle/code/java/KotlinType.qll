/**
 * Provides classes and predicates for working with Kotlin types.
 */

import java

class KotlinType extends Element, @kt_type {
}

class KotlinNullableType extends KotlinType, @kt_nullable_type {
  override string toString() {
    exists(ClassOrInterface ci |
           kt_nullable_types(this, ci) and
           result = "Kotlin nullable " + ci.toString())
  }
  override string getAPrimaryQlClass() { result = "KotlinNullableType" }
}

class KotlinNotnullType extends KotlinType, @kt_notnull_type {
  override string toString() {
    exists(ClassOrInterface ci |
           kt_notnull_types(this, ci) and
           result = "Kotlin not-null " + ci.toString())
  }
  override string getAPrimaryQlClass() { result = "KotlinNotnullType" }
}

class KotlinTypeAlias extends Element, @kt_type_alias {
  override string getAPrimaryQlClass() { result = "KotlinTypeAlias" }

  KotlinType getKotlinType() {
    kt_type_alias(this, _, result)
  }
}
