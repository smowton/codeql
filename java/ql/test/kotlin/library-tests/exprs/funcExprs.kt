fun functionExpression0a(f: () -> Int) { }
fun functionExpression0b(f: () -> Any?) { }
fun functionExpression0c(f: () -> Any) { }
fun functionExpression1a(x: Int, f: (Int) -> Int) { }
fun functionExpression1b(x: Int, f: (Any?) -> Any?) { }
fun functionExpression2(x: Int, f: (Int, Int) -> Int) { }
fun functionExpression3(x: Int, f: Int.(Int) -> Int) { }
fun functionExpression4(x: Int, f: (Int) -> ((Int) -> Double)) { }

fun functionExpression22(x: Int, f: (Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int) -> Unit) {}
fun functionExpression23(x: Int, f: (Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int, Int) -> String) {}

fun call() {
    functionExpression0a { -> 5 }
    functionExpression0b { -> 5 }
    functionExpression0c { -> 5 }
    functionExpression1a(5) { a -> 5 }
    functionExpression1a(5) { it }
    functionExpression1a(5, fun(_:Int) = 5)
    functionExpression1a(5, MyLambda())
    functionExpression1b(5) { a -> a}
    functionExpression2(5, fun(_: Int, _: Int) = 5)
    functionExpression2(5) { _, _ -> 5 }
    functionExpression3(5) { a -> this + a }
    functionExpression4(5) { a -> ( { b -> 5.0} ) }

    functionExpression22(5) {a0,a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11,a12,a13,a14,a15,a16,a17,a18,a19,a20,a21 -> 5}
    functionExpression23(5) {a0,a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11,a12,a13,a14,a15,a16,a17,a18,a19,a20,a21,a22 -> ""}
}

class MyLambda: (Int) -> Int {
    override operator fun invoke(x: Int): Int = 5
}