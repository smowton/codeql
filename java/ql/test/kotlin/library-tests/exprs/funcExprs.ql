import java

private string getLambdaBody(LambdaExpr le) {
  le.hasExprBody() and result = "expr body"
  or
  le.hasStmtBody() and result = "stmt body"
}

query predicate lambdaExpr(LambdaExpr le, string body, Method m, string signature, AnonymousClass an) {
  getLambdaBody(le) = body and
  le.asMethod() = m and
  signature = m.getSignature() and
  le.getAnonymousClass() = an
}

query predicate memberRefExprs(MemberRefExpr e, Method m, string signature, AnonymousClass an) {
  e.asMethod() = m and
  signature = m.getSignature() and
  e.getAnonymousClass() = an
}
