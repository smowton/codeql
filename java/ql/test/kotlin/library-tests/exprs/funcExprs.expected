| funcExprs.kt:14:26:14:33 | ...->... | stmt body | funcExprs.kt:14:26:14:33 | invoke | invoke() | funcExprs.kt:14:26:14:33 | new Function0<Integer>(...) { ... } |
| funcExprs.kt:15:26:15:33 | ...->... | stmt body | funcExprs.kt:15:26:15:33 | invoke | invoke() | funcExprs.kt:15:26:15:33 | new Function0<Object>(...) { ... } |
| funcExprs.kt:16:26:16:33 | ...->... | stmt body | funcExprs.kt:16:26:16:33 | invoke | invoke() | funcExprs.kt:16:26:16:33 | new Function0<Object>(...) { ... } |
| funcExprs.kt:17:29:17:38 | ...->... | stmt body | funcExprs.kt:17:29:17:38 | invoke | invoke(int) | funcExprs.kt:17:29:17:38 | new Function1<Integer,Integer>(...) { ... } |
| funcExprs.kt:18:29:18:34 | ...->... | stmt body | funcExprs.kt:18:29:18:34 | invoke | invoke(int) | funcExprs.kt:18:29:18:34 | new Function1<Integer,Integer>(...) { ... } |
| funcExprs.kt:19:29:19:42 | ...->... | stmt body | funcExprs.kt:19:29:19:42 | invoke | invoke(int) | funcExprs.kt:19:29:19:42 | new Function1<Integer,Integer>(...) { ... } |
| funcExprs.kt:21:29:21:37 | ...->... | stmt body | funcExprs.kt:21:29:21:37 | invoke | invoke(java.lang.Object) | funcExprs.kt:21:29:21:37 | new Function1<Object,Object>(...) { ... } |
| funcExprs.kt:22:28:22:50 | ...->... | stmt body | funcExprs.kt:22:28:22:50 | invoke | invoke(int,int) | funcExprs.kt:22:28:22:50 | new Function2<Integer,Integer,Integer>(...) { ... } |
| funcExprs.kt:23:28:23:40 | ...->... | stmt body | funcExprs.kt:23:28:23:40 | invoke | invoke(int,int) | funcExprs.kt:23:28:23:40 | new Function2<Integer,Integer,Integer>(...) { ... } |
| funcExprs.kt:24:28:24:44 | ...->... | stmt body | funcExprs.kt:24:28:24:44 | invoke | invoke(int,int) | funcExprs.kt:24:28:24:44 | new Function2<Integer,Integer,Integer>(...) { ... } |
| funcExprs.kt:25:28:25:51 | ...->... | stmt body | funcExprs.kt:25:28:25:51 | invoke | invoke(int) | funcExprs.kt:25:28:25:51 | new Function1<Integer,Function1<Integer,Double>>(...) { ... } |
| funcExprs.kt:25:37:25:47 | ...->... | stmt body | funcExprs.kt:25:37:25:47 | invoke | invoke(int) | funcExprs.kt:25:37:25:47 | new Function1<Integer,Double>(...) { ... } |
| funcExprs.kt:27:29:27:112 | ...->... | stmt body | funcExprs.kt:27:29:27:112 | invoke | invoke(int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int) | funcExprs.kt:27:29:27:112 | new Function22<Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Integer,Unit>(...) { ... } |
| funcExprs.kt:28:29:28:117 | ...->... | stmt body | funcExprs.kt:28:29:28:117 | invoke | invoke(int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int,int) | funcExprs.kt:28:29:28:117 | new FunctionN<String>(...) { ... } |
| funcExprs.kt:28:29:28:117 | ...->... | stmt body | funcExprs.kt:28:29:28:117 | invoke | invoke(java.lang.Object[]) | funcExprs.kt:28:29:28:117 | new FunctionN<String>(...) { ... } |
